go/tools/builders/nogo_change.go

```diff
@@ -0,0 +1,344 @@
+}
+
+// NogoChange represents a collection of file edits.
+type NogoChange struct {
+	FileToEdits map[string]NogoFileEdits // File path as the key, analyzer-to-edits mapping as the value
+}
+
+// newChange creates a new NogoChange object.
+func newChange() *NogoChange {
+	return &NogoChange{
+		FileToEdits: make(map[string]NogoFileEdits),
+	}
+}
+
+func (e NogoEdit) String() string {
+	return fmt.Sprintf("{Start:%d,End:%d,New:%q}", e.Start, e.End, e.New)
+}
+
+// sortEdits orders a slice of NogoEdits by (start, end) offset.
+// This ordering puts insertions (end = start) before deletions
+// (end > start) at the same point, but uses a stable sort to preserve
+// the order of multiple insertions at the same point.
+// (applyEditsBytes detects multiple deletions at the same point as an error.)
+func sortEdits(edits []NogoEdit) {
+	sort.Stable(editsSort(edits))
+}
+
+type editsSort []NogoEdit
+
+func (a editsSort) Len() int { return len(a) }
+func (a editsSort) Less(i, j int) bool {
+	if cmp := a[i].Start - a[j].Start; cmp != 0 {
+		return cmp < 0
+	}
+	return a[i].End < a[j].End
+}
+func (a editsSort) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+
+// validateBytes checks that edits are consistent with the src byte slice,
+// and returns the size of the patched output. It may return a different slice if edits are sorted.
+func validateBytes(src []byte, edits []NogoEdit) ([]NogoEdit, int, error) {
+	if !sort.IsSorted(editsSort(edits)) {
+		edits = append([]NogoEdit(nil), edits...)
+		sortEdits(edits)
+	}
+
+	size := len(src)
+	lastEnd := 0
+	for _, edit := range edits {
+		if !(0 <= edit.Start && edit.Start <= edit.End && edit.End <= len(src)) {
+			return nil, 0, fmt.Errorf("the fix has an out-of-bounds edit with start=%d, end=%d", edit.Start, edit.End)
+		}
+		if edit.Start < lastEnd {
+			return nil, 0, fmt.Errorf("the fix has an edit with start=%d, which overlaps with a previous edit with end=%d", edit.Start, lastEnd)
+		}
+		size += len(edit.New) + edit.Start - edit.End
+		lastEnd = edit.End
+	}
+
+	return edits, size, nil
+}
+
+// applyEditsBytes applies a sequence of NogoEdits to the src byte slice and returns the result.
+// Edits are applied in order of start offset; edits with the same start offset are applied in the order they were provided.
+// applyEditsBytes returns an error if any edit is out of bounds, or if any pair of edits is overlapping.
+func applyEditsBytes(src []byte, edits []NogoEdit) ([]byte, error) {
+	edits, size, err := validateBytes(src, edits)
+	if err != nil {
+		return nil, err
+	}
+
+	// Apply the edits.
+	out := make([]byte, 0, size)
+	lastEnd := 0
+	for _, edit := range edits {
+		if lastEnd < edit.Start {
+			out = append(out, src[lastEnd:edit.Start]...)
+		}
+		out = append(out, edit.New...)
+		lastEnd = edit.End
+	}
+	out = append(out, src[lastEnd:]...)
+
+	if len(out) != size {
+		return nil, fmt.Errorf("applyEditsBytes: unexpected output size, got %d, want %d", len(out), size)
```

Comment:
How can this happen?

Reply:
it cannot happen. 
code greatly simplified now.